--- squashfs3.0/squashfs-tools/squashfs_fs.h~	2006-03-15 22:36:21.000000000 +0100
+++ squashfs3.0/squashfs-tools/squashfs_fs.h	2006-03-18 21:28:36.000000000 +0100
@@ -26,8 +26,8 @@
 
 #define SQUASHFS_MAJOR			3
 #define SQUASHFS_MINOR			0
-#define SQUASHFS_MAGIC			0x73717368
-#define SQUASHFS_MAGIC_SWAP		0x68737173
+#define SQUASHFS_MAGIC			0x7371736c
+#define SQUASHFS_MAGIC_SWAP		0x6c737173
 #define SQUASHFS_START			0
 
 /* size of metadata (inode and directory) blocks */
